python/pyarrow/_parquet.pyx

```diff
@@ -505,6 +506,204 @@ cdef class ColumnChunkMetaData(_Weakrefable):
+        self.descending = descending
+        self.nulls_first = nulls_first
+
+    @classmethod
+    def from_sort_order(cls, Schema schema, sort_keys, null_placement='at_end'):
+        """
+        Create a tuple of SortingColumn objects from the same arguments as
+        :class:`pyarrow.compute.SortOptions`.
+
+        Parameters
+        ----------
+        schema : Schema
+            Schema of the input data.
+        sort_keys : Sequence of (name, order) tuples
+            Names of field/column keys (str) to sort the input on,
+            along with the order each field/column is sorted in.
+            Accepted values for `order` are "ascending", "descending".
+        null_placement : {'at_start', 'at_end'}, default 'at_end'
+            Where null values should appear in the sort order.
+
+        Returns
+        -------
+        sorting_columns : tuple of SortingColumn
+        """
+        if null_placement == 'at_start':
+            nulls_first = True
+        elif null_placement == 'at_end':
+            nulls_first = False
+        else:
+            raise ValueError('null_placement must be "at_start" or "at_end"')
+
+        col_map = _name_to_index_map(schema)
+
+        sorting_columns = []
+
+        for sort_key in sort_keys:
+            if isinstance(sort_key, str):
+                name = sort_key
+                descending = False
+            elif (isinstance(sort_key, tuple) and len(sort_key) == 2 and
+                    isinstance(sort_key[0], str) and
+                    isinstance(sort_key[1], str)):
+                name, descending = sort_key
+                if descending == "descending":
+                    descending = True
+                elif descending == "ascending":
+                    descending = False
+                else:
+                    raise ValueError("Invalid sort key direction: {0}"
+                                     .format(descending))
+            else:
+                raise ValueError("Invalid sort key: {0}".format(sort_key))
+
+            try:
+                column_index = col_map[name]
+            except KeyError:
+                raise ValueError("Sort key name '{0}' not found in schema:\n{1}"
+                                 .format(name, schema))
+
+            sorting_columns.append(
+                cls(column_index, descending=descending, nulls_first=nulls_first)
+            )
+
+        return tuple(sorting_columns)
+
+    @staticmethod
+    def to_sort_order(Schema schema, sorting_columns):
+        """
+        Convert a tuple of SortingColumn objects to the same format as
+        :class:`pyarrow.compute.SortOptions`.
+
+        Parameters
+        ----------
+        schema : Schema
+            Schema of the input data.
+        sorting_columns : tuple of SortingColumn
+            Columns to sort the input on.
+
+        Returns
+        -------
+        sort_keys : tuple of (name, order) tuples
+        null_placement : {'at_start', 'at_end'}
```

Comment:
This is due to inconsistencies between what is possible to store in Parquet and what is possible to (simply) generate in Arrow. In the Parquet spec each column of the `sorting_columns` can indeed have a different `null_placement`, however, it isn't possible to express that in Arrow with a single sort since [SortOptions](https://arrow.apache.org/docs/python/generated/pyarrow.compute.SortOptions.html#pyarrow.compute.SortOptions) has a global `null_placement`.

Reading the code again it isn't obvious that this is the case and I think that `to/from_sort_order` should probably be `to/from_sort_options` and perhaps these methods should actually return a `SortOptions` object? The downside to that is that the `SortOptions` class actually isn't very ergonomic as it can't be used directly with `Table/RecordBatch/Dataset.sort_by` - maybe these methods could take an optional `options: SortOptions` kwarg like some other methods do which would ease this pain?

Reply:
Could we rename the methods to be `to/from_ordering` to match the naming in the C++ where `SortOrder` is a part of the ordering class through `sort_keys`:

https://github.com/apache/arrow/blob/0479f8532d5ace54cf554e2e60aa621f06536091/cpp/src/arrow/compute/ordering.h#L63-L65

Reply:
https://github.com/apache/arrow/pull/37665/commits/44dff9dff6f736bc97e829951fa5271dc073aa18